dds/DCPS/DiscoveryBase.h

```diff
@@ -1051,170 +1121,68 @@ namespace OpenDDS {
-                DDS::Security::DatareaderCryptoHandle drch = iter->second;
-                DDS::Security::DatawriterCryptoHandle dwch =
-                  generate_remote_matched_writer_crypto_handle(writer_participant, drch);
-                remote_writer_crypto_handles_[writer] = dwch;
-                DatawriterCryptoTokenSeqMap::iterator t_iter =
-                  pending_remote_writer_crypto_tokens_.find(writer);
-                if (t_iter != pending_remote_writer_crypto_tokens_.end()) {
-                  DDS::Security::SecurityException se;
-                  if (!keyexg->set_remote_datawriter_crypto_tokens(iter->second, dwch, t_iter->second, se)) {
-                    ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) ERROR: ")
-                      ACE_TEXT("(%P|%t) ERROR: DiscoveryBase::match() - ")
-                      ACE_TEXT("Unable to set pending remote datawriter crypto tokens with ")
-                      ACE_TEXT("crypto key exchange plugin. Security Exception[%d.%d]: %C\n"),
-                        se.code, se.minor_code, se.message.in()));
-                  }
-                  pending_remote_writer_crypto_tokens_.erase(t_iter);
-                }
-                EndpointSecurityAttributesMap::const_iterator s_iter =
-                  local_reader_security_attribs_.find(reader);
-                // Yes, this is different for remote datawriters than readers (see 8.8.9.3 vs 8.8.9.2)
-                if (s_iter != local_reader_security_attribs_.end() && s_iter->second.is_submessage_protected) {
-                  create_and_send_datareader_crypto_tokens(drch, reader, dwch, writer);
-                }
+          MatchingData md;
+          ACE_GUARD(ACE_Thread_Mutex, g1, matching_data_buffer_lock_);
+
+          // if the type object is not in cache, send RPC request
+          md.dwQos = dwQos;
+          md.pubQos = pubQos;
+          md.drQos = drQos;
+          md.subQos = subQos;
+          md.wTls = wTls;
+          md.rTls = rTls;
+          md.writer_type_info = writer_type_info;
+          md.reader_type_info = reader_type_info;
+          md.cfProp = cfProp;
+          md.writer = writer;
+          md.reader = reader;
+          md.time_added_to_map = MonotonicTimePoint::now();
+
+          // TLS_TODO: rpc communication needs to be tested when XTypes are ready
+          if (!writer_local) {
+            if (type_lookup_service_ && !type_lookup_service_->type_object_in_cache(writer_type_info->minimal.typeid_with_size.type_id)) {
+              XTypes::TypeIdentifierSeq type_ids;
+              type_ids.append(writer_type_info->minimal.typeid_with_size.type_id);
+              md.rpc_sequence_number = ++(type_lookup_service_->rpc_sequence_number_);
+              MatchingDataIter md_it = matching_data_buffer_.find(MatchingPair(writer, reader));
+              if (md_it != matching_data_buffer_.end()) {
+                // TLS_TODO: is this scenario possible?
```

Comment:
I'm not sure I see the problem. The data has to get into `matching_data_buffer_`, so there would be a point where it's not there.

Reply:
I am just wondering if we should expect a situation when the same two endpoints try to match more than once.

Reply:
I'm not sure. It sounds like it shouldn't be possible, but I don't think how you're handling it is going to hurt anything.